simtfl/node.py

```diff
@@ -78,12 +105,72 @@ def run(self):
         implementation.
         """
         while True:
-            while len(self.mailbox) > 0:
-                (sender, message) = self.mailbox.popleft()
+            while len(self._mailbox) > 0:
+                (sender, message) = self._mailbox.popleft()
                 print(f"T{self.env.now:5d}: handling  {sender:2d} -> {self.ident:2d}: {message}")
                 yield from self.handle(sender, message)
 
             # This naive implementation is fine because we have no actual
             # concurrency.
-            self.wakeup = self.env.event()
-            yield self.wakeup
+            self._wakeup = self.env.event()
+            yield self._wakeup
+
+
+__all__ = ['PassiveNode', 'SequentialNode']
+
+from simpy import Environment
+import unittest
+
+from .message import PayloadMessage
+from .network import Network
+
+
+class PassiveReceiverTestNode(PassiveNode):
+    def __init__(self):
+        super().__init__()
+        self.received = deque()
+
+    def handle(self, sender, message):
+        self.received.append((sender, message, self.env.now))
+        yield self.env.timeout(3)
+
+
+class SequentialReceiverTestNode(SequentialNode):
+    def __init__(self):
+        super().__init__()
+        self.received = deque()
+
+    def handle(self, sender, message):
+        self.received.append((sender, message, self.env.now))
+        yield self.env.timeout(3)
+
+
+class SenderTestNode(PassiveNode):
+    def run(self):
+        for i in range(3):
+            yield from self.send(0, PayloadMessage(i))
+            yield self.env.timeout(1)
+
+
+class TestFramework(unittest.TestCase):
+    def _test_node(self, receiver_node, expected):
+        network = Network(Environment())
+        network.add_node(receiver_node)
+        network.add_node(SenderTestNode())
+        network.run_all()
+
+        self.assertEqual(list(network.node(0).received), expected)
+
+    def test_passive_node(self):
+        self._test_node(PassiveReceiverTestNode(), [
+            (1, PayloadMessage(0), 1),
+            (1, PayloadMessage(1), 2),
+            (1, PayloadMessage(2), 3),
+        ])
+
+    def test_sequential_node(self):
+        self._test_node(SequentialReceiverTestNode(), [
+            (1, PayloadMessage(0), 1),
+            (1, PayloadMessage(1), 4),
+            (1, PayloadMessage(2), 7),
```

Comment:
Here the messages are being sent at times 0, 1, 2, received at times 1, 2, 3, and handled sequentially. So the first is handled when it is received at time 1, but the second and third can only be handled 3 and 6 time units later.

Reply:
My confusion in the above comment doesn't seem to extend here, so long as I assume the first message is "received" at time 1, then I understand why the remaining messages are received at 4 and 7 due to `SequentialNode` serializing them (and doing an `env.timeout` after receiving each).

So I think my confusion is primarily around sending and timeouts and how that is modeling network propagation delay (plus sender processing time?). In the case of the receiver I think of the `env.timeout` calls as representing "time to process a message".